debian/drush.1
.\"                                      Hey, EMACS: -*- nroff -*-
'\" t -*- coding: UTF-8 -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH DRUSH 1 "June 11, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
drush \- shell scripting interface for Drupal
.SH SYNOPSIS
.B drush
.RI [ options ] " command" ...
.SH DESCRIPTION
.PP
\fBdrush\fP is a veritable Swiss Army knife designed to make life
easier for those of us who spend some of our working hours hacking
away at the command prompt.
.PP
\fBDrush\fP core ships with lots of useful commands for interacting with
code like modules/themes/profiles. Similarly, it runs update.php,
executes sql queries and DB migrations, and misc utilities like run
cron or clear cache.
.SH OPTIONS
.TP
.BI \-r\  path ,\ \-\-root= path
Drupal root directory to use (default: current directory)             
.TP
.BI \-l\  uri ,\ \-\-uri= uri
URI of the drupal site to use (only needed in multisite environments) 
.TP
.BI \-v\  level ,\ \-\-verbose= level
Display extra information about the command.                          
.TP
.B \-q, \-\-quiet
Hide all output
.TP
.B \-d, \-\-debug
Display even more information, including internal messages.           
.TP
.B \-y, \-\-yes
Assume 'yes' as answer to all prompts                                 
.TP
.B \-s, \-\-simulate
Simulate all relevant actions (don't actually change the system)
.TP
.B \-i, \-\-include
A list of paths to search for drush commands
.TP
.BI \-c\  file ,\ \-\-config= file
Specify a config file to use. See 
.IR example.drushrc.php                 
.TP
.BI \-u\  user ,\ \-\-user= user
Specify a user to login with. May be a name or a number.              
.TP
.B \-b, \-\-backend
Hide all output and return structured data (internal use only).       
.TP
.B \-p, \-\-pipe
Emit a compact representation of the command for scripting.           
.PP
If you get tired of typing options all the time, you can add them to
your drush.php alias or create a drushrc.php file. These provide
additional options for your drush call. They provide great flexibility
for a multi-site installation, for example. See example.drushrc.php.
.SH COMMANDS
Drush is all based around "commands" that are usually defined in
various modules. This manual only lists the builtin commands shipped
with the Drush core, but other modules may extend this list. See the
.B help
command to see the up to date list. The list of available comands may
also vary based on the state of the Drupal accessed by
.B drush
(if it's installed, offline, working or not, which modules are
installed, etc).
.TP
.B help
Print the help message. Use --filter to limit command list to one
command file (e.g. --filter=pm).
.TP
.B cron
Run all cron hooks.
.TP
.B updatedb
Execute the update.php process from the command line
.TP
.B status
Provides a birds-eye view of the current Drupal installation, if any.
.TP
.B script
Run php script(s).
.TP
.B cache clear
Clear all caches.
.TP
.B watchdog show
Shows recent watchdog log messages. Optionally filter for a specific type.
.TP
.B watchdog delete
Delete all messages or only those of a specified type.
.TP
.B sync
Rsync the Drupal tree to/from another server using ssh.
.TP
.B eval
Evaluate arbitrary php code after bootstrapping Drupal.
.TP
.B enable
Enable one or more modules.
.TP
.B disable
Disable one or more modules.
.TP
.B uninstall
Uninstall one or more modules.
.TP
.B statusmodules
Show module enabled/disabled status
.TP
.B refresh
Refresh update status information
.TP
.B updatecode
Update your project code
.TP
.B update
Update your project code and apply any database updates required (update.php)
.TP
.B info
Release information for a project
.TP
.B dl
Download core Drupal and projects like CCK, Zen, etc.
.TP
.B test mail
Run all tests and mail the results to your team.
.TP
.B test clean
Delete leftover tables and files from prior test runs.
.TP
.B sql conf
Print database connection details.
.TP
.B sql connect
A string for connecting to the DB.
.TP
.B sql dump
Exports the Drupal DB as SQL using mysqldump.
.TP
.B sql query
Execute a query against the site database.
.TP
.B sql load
Copy source database to target database.
.TP
.B sql cli
Open a SQL command-line interface using Drupal's credentials.
.SH FILES
.IR /etc/drush/drushrc.php
.br
.I /usr/share/doc/drush/examples/example.drushrc.php
.SH SEE ALSO
.BR cvs (1),
.BR svn (1),
.BR wget (1).
.SH AUTHOR
Drush was originally developed by Arto for Drupal 4.7 (this alpha code
can still be found in the DRUPAL-4-7 branch). In May 2007, it was
partly rewritten and redesigned for Drupal 5 by frando. Since 2008,
the module has been maintained by Moshe Weitzman, Owen Barton and
Adrian Rossouw.
.PP
This manual page was written by Antoine Beaupré <anarcat@koumbit.org>,
for the Debian project (but may be used by others).


debian/changelog
drush (2.0-4~koumbit50+1) stable; urgency=low

  * backport to lenny
  * add VCS-* fields to debian/control file

 -- Antoine Beaupré <anarcat@koumbit.org>  Wed, 17 Jun 2009 21:07:18 -0400

drush (2.0-3) unstable; urgency=low

  * use dh_prep instead of dh_clean to silence warnings on sid. this will
    now fail to build on lenny.
  * cleanup comments in watch and rules files
  * update to standards 3.8.1

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 17:10:59 -0400

drush (2.0-2) unstable; urgency=medium

  * don't use the bash wrapper because it breaks in cron

 -- Antoine Beaupré <anarcat@koumbit.org>  Sun, 14 Jun 2009 16:47:43 -0400

drush (2.0-1) unstable; urgency=low

  * Initial release (Closes: #532723)
  * Write a manpage
  * Includes the table.inc file from PEAR's Console_Table

 -- Antoine Beaupré <anarcat@koumbit.org>  Thu, 11 Jun 2009 21:56:06 -0400



drush.php
#!/usr/bin/env php
<?php
// $Id: drush.php,v 1.67 2009/06/04 03:56:07 grugnog Exp $

/**
 * @file
 * drush is a PHP script implementing a command line shell for Drupal.
 *
 * @requires PHP CLI 5.2.0, or newer.
 */

// Terminate immediately unless invoked as a command line script
if (!drush_verify_cli()) {
  die('drush.php is designed to run via the command line.');
}

// Check supported version of PHP.
define('DRUSH_MINIMUM_PHP', '5.2.0');
if (version_compare(phpversion(), DRUSH_MINIMUM_PHP) < 0) {
  die('Your PHP installation is too old. Drush requires at least PHP ' . DRUSH_MINIMUM_PHP . "\n");
}

define('DRUSH_BASE_PATH', dirname(__FILE__));
define('DRUSH_COMMAND', $GLOBALS['argv'][0]);
define('DRUSH_REQUEST_TIME', microtime(TRUE));

require_once DRUSH_BASE_PATH . '/includes/environment.inc';
require_once DRUSH_BASE_PATH . '/includes/command.inc';
require_once DRUSH_BASE_PATH . '/includes/drush.inc';
require_once DRUSH_BASE_PATH . '/includes/backend.inc';
require_once DRUSH_BASE_PATH . '/includes/context.inc';

drush_set_context('argc', $GLOBALS['argc']);
drush_set_context('argv', $GLOBALS['argv']);

exit(drush_main());

/**
 * Verify that we are running PHP through the command line interface.
 *
 * This function is useful for making sure that code cannot be run via the web server,
 * such as a function that needs to write files to which the web server should not have
 * access to.
 *
 * @return
 *   A boolean value that is true when PHP is being run through the command line,
 *   and false if being run through cgi or mod_php.
 */
function drush_verify_cli() {
  return (php_sapi_name() == 'cli' || (is_numeric($_SERVER['argc']) && $_SERVER['argc'] > 0));
}

/**
 * The main Drush function.
 *
 * - Parses the command line arguments, configuration files and environment.
 * - Prepares and executes a Drupal bootstrap, if possible,
 * - Dispatches the given command.
 *
 * @return
 *   Whatever the given command returns.
 */
function drush_main() {
  $phases = _drush_bootstrap_phases();

  $return = '';
  $command_found = FALSE;

  foreach ($phases as $phase) {
    if (drush_bootstrap($phase)) {
      $command = drush_parse_command();
      if (is_array($command)) {
        if ($command['bootstrap'] == $phase && empty($command['bootstrap_errors'])) {
          drush_log(dt("Found command: !command", array('!command' => $command['command'])), 'bootstrap');
          $command_found = TRUE;
          // Dispatch the command(s).
          $return = drush_dispatch($command);
          break;
        }
      }
    }
    else {
      break;
    }
  }

  if (!$command_found) {
    // If we reach this point, we have not found either a valid or matching command.
    $args = implode(' ', drush_get_arguments());
    $drush_command = array_pop(explode('/', DRUSH_COMMAND));
    if (isset($command) && is_array($command)) {
      foreach ($command['bootstrap_errors'] as $key => $error) {
        drush_set_error($key, $error); 
      }
      drush_set_error('DRUSH_COMMAND_NOT_EXECUTABLE', dt("The command '!drush_command !args' could not be executed.", array('!drush_command' => $drush_command, '!args' => $args)));
    }
    elseif (!empty($args)) {
      drush_set_error('DRUSH_COMMAND_NOT_FOUND', dt("The command '!drush_command !args' could not be found.", array('!drush_command' => $drush_command, '!args' => $args)));
    }
    else {
      // This can occur if we get an error during _drush_bootstrap_drush_validate();
      drush_set_error('DRUSH_COULD_NOT_EXECUTE', dt("Drush could not execute."));
    }
  }

  // We set this context to let the shutdown function know we reached the end of drush_main();
  drush_set_context("DRUSH_EXECUTION_COMPLETED", TRUE);

  // After this point the drush_shutdown function will run,
  // exiting with the correct exit code.
  return $return;
}

/**
 * Shutdown function for use while Drupal is bootstrapping and to return any
 * registered errors.
 *
 * The shutdown command checks whether certain options are set to reliably
 * detect and log some common Drupal initialization errors.
 *
 * If the command is being executed with the --backend option, the script
 * will return a json string containing the options and log information
 * used by the script.
 * 
 * The command will exit with '1' if it was succesfully executed, and the 
 * result of drush_get_error() if it wasn't.
 */
function drush_shutdown() {
  // Mysteriously make $user available during sess_write(). Avoids a NOTICE.
  global $user; 

  if (!drush_get_context('DRUSH_EXECUTION_COMPLETED', FALSE)) {
    // We did not reach the end of the drush_main function, 
    // this generally means somewhere in the code a call to exit(),
    // was made. We catch this, so that we can trigger an error in 
    // those cases.
    drush_set_error("DRUSH_NOT_COMPLETED", dt("Drush command could not be completed."));
  }

  $phase = drush_get_context('DRUSH_BOOTSTRAP_PHASE');
  if (drush_get_context('DRUSH_BOOTSTRAPPING')) {
    switch ($phase) {
      case DRUSH_BOOTSTRAP_DRUPAL_FULL :
        ob_end_clean();
        _drush_log_drupal_messages();
        drush_set_error('DRUSH_DRUPAL_BOOTSTRAP_ERROR');
        break;
    }
  }

  if (drush_get_context('DRUSH_BACKEND')) {
    drush_backend_output();
  }
  elseif (drush_get_context('DRUSH_QUIET')) {
    ob_end_clean();
  }
  
  // If we are in pipe mode, emit the compact representation of the command, if available.
  if (drush_get_context('DRUSH_PIPE')) {
    drush_pipe_output();
  }
  
  exit((drush_get_error()) ? DRUSH_FRAMEWORK_ERROR : DRUSH_SUCCESS);
}

/**
 * Log the given user in to a bootstrapped Drupal site.
 *
 * @param mixed
 *   Numeric user id or user name.
 *
 * @return boolean
 *   TRUE if user was logged in, otherwise FALSE.
 */
function drush_drupal_login($drush_user) {
  global $user;
  if (drush_drupal_major_version() >= 7) {
    $user = is_numeric($drush_user) ? user_load($drush_user) : user_load_by_name($drush_user);
  }
  else {
    $user = user_load(is_numeric($drush_user) ? array('uid' => $drush_user) : array('name' => $drush_user));
  }

  if (empty($user)) {
    if (is_numeric($drush_user)) {
      $message = dt('Could not login with user ID #%user.', array('%user' => $drush_user));
    }
    else {
      $message = dt('Could not login with user account `%user\'.', array('%user' => $drush_user));
    }
    return drush_set_error('DRUPAL_USER_LOGIN_FAILED', $message);
  }

  return TRUE;
}

